File: CoreSkills/Graphs/primsMST.cpp
#include <unordered_map>
#include <vector>
#include <tuple>
#include <unordered_set>
#include <queue>

class Solution {
public:
    int minimumSpanningTree(std::vector<std::vector<int>>& edges, int n) {
        /*
        - implement Prim's MST algo
        - An MST is the tree (set of edges) that connects all vertices in an undirected graph for the minimum possible cost
        - An MST is acylic
        - Algo: starting from an arbitrary vertex, add to the MST the smallest cost edge to an unvisited node
        - Similar to Dijkstra's and bfs: use a minheap to keep track of edges and a set to keep track of visited nodes
        - add outgoing edges to the minheap, process the cheapest edge to an unvisited node
        - input: weighted undirected graph
        - edges: a vector of vectors (u,v,w) where u and v are vertices and w is weight
        - n: number of vertices
        - Output: total weight of the MST (if the graph is not connected return -1)

        Input:
        n = 5
        edges = [[0,1,10], [0,2,3], [1,3,2], [2,1,4], [2,3,8], [2,4,2], [3,4,5]]

        Output:
        11
        */

        // build adjacency list
        std::unordered_map<int, std::vector<std::tuple<int,int>>> graph; // key: vertex, value: vector of tuples(weight, dst vertex)
        for (int i = 0; i < edges.size(); i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            int w = edges[i][2];
            // add the edge to the adjacency list of u and v
            graph[u].push_back(std::make_tuple(w, v));
            graph[v].push_back(std::make_tuple(w, u));
        }

        // setup prim's (start at vertex 0)
        std::unordered_set<int> visited;
        std::priority_queue<std::tuple<int,int>> minheap; // maxheap by default, push edges with negative weight to achieve a minheap
        int totalweight = 0;
        visited.insert(0);
        for (std::tuple<int,int> edge: graph[0]) {
            int w = get<0>(edge);
            int dst = get<1>(edge);
            minheap.push(std::make_tuple(-1 * w, dst));
        }

        // execute prim's
        while (!minheap.empty()) {
            auto edge = minheap.top();
            minheap.pop();
            int w = get<0>(edge) * -1; // make the weight a positive number
            int dst = get<1>(edge);

            // add edge to MST if dst is unvisited
            if (visited.find(dst) == visited.end()) {
                visited.insert(dst);
                totalweight = totalweight + w;
            }

            // add outgoing edges from dst to unvisited vertices to the minheap
            for (std::tuple<int,int> e: graph[dst]) {
                int weight = get<0>(e);
                int v = get<1>(e);
                if (visited.find(v) == visited.end()) {
                    minheap.push(std::make_tuple(-1*weight, v));
                }
            }
        }

        // return totalweight or -1 if not all vertices were visited
        if (visited.size() == n) {
            return totalweight;
        }
        return -1;

        // Time complexity is O(ElogE) because in the worst case we have to push and pop E edges from the graph.
        // pushing and popping is worst case O(logE) so total is O(ElogE)
    }
};

File: CoreSkills/Graphs/primsMST.py
import heapq

class Solution:
    def minimumSpanningTree(self, n: int, edges: list[list[int]]) -> int:
        """
        - Implement prim's algorithm (minimum spanning tree)
        - What is a minimum spanning tree? An acyclic tree that connects all vertices in the graph for the 
        minimum cost (undirected graph)
        - Use V - 1 edges to connect V nodes
        - greedy algorithm: add the min edge to an unvisited node at each step (growing the tree)
        - find MST using Prim's and return total weight
        - return -1 if graph is unconnected

        Input:
        n = 5 (number of vertices)
        edges = [[0,1,10], [0,2,3], [1,3,2], [2,1,4], [2,3,8], [2,4,2], [3,4,5]] (u, v, w) (edge from u to v with cost w)

        Output:
        11
        """
        totalweight = 0
        
        # 1. Build adjacency list
        graph = {}
        for i in range(n):
            graph[i] = []
        
        # add edges to graph
        for e in edges:
            u = e[0]
            v = e[1] 
            w = e[2]
            # undirected - edge goes both ways, add edge to adjacency list of u and v
            graph[u].append([w, v]) # [weight, v]
            graph[v].append([w, u])
        
        # 2. setup Prim's
        visited = set()
        minheap = [] # contain tuples (weight, vertex)
        # start at 0
        visited.add(0)
        for e in graph[0]:
            weight = e[0]
            v = e[1]
            heapq.heappush(minheap, (weight, v))

        # 3. execute Prim's
        while len(minheap) > 0:
            e = heapq.heappop(minheap)
            weight = e[0]
            v = e[1]

            # updating the total weight
            if v not in visited:
                visited.add(v)
                totalweight += weight

            # add outgoing edges
            for edge in graph[v]:
                cost = edge[0]
                vertex = edge[1]
                if vertex not in visited:
                    heapq.heappush(minheap, (cost, vertex))
        
        # 4. return total weight or -1 if graph is unconnected
        if len(visited) != n:
            return -1
        
        return totalweight

"""
Time complexity: O(ElogE) because we have to process E edges by pushing and popping from the minheap which takes logE.
"""


        